File: src/main/java/com/algorand/algosdk/v2/client/indexer/LookupAssetTransactions.java
package com.algorand.algosdk.v2.client.indexer;

import com.algorand.algosdk.v2.client.common.Client;
import com.algorand.algosdk.v2.client.common.Query;
import com.algorand.algosdk.v2.client.common.QueryData;
import com.algorand.algosdk.v2.client.common.Response;
import com.algorand.algosdk.v2.client.model.TransactionsResponse;


/**
 * Lookup transactions for an asset. /assets/{asset-id}/transactions 
 */
public class LookupAssetTransactions extends Query {
	private String address;
	private String addressRole;
	private String afterTime;
	private Long assetId;
	private String beforeTime;
	private Long currencyGreaterThan;
	private Long currencyLessThan;
	private Boolean excludeCloseTo;
	private Long limit;
	private Long maxRound;
	private Long minRound;
	private String next;
	private String notePrefix;
	private Long round;
	private String sigType;
	private String txId;
	private String txType;


	public LookupAssetTransactions(Client client, Long assetId) {
		super(client, "get");
		this.assetId = assetId;
	}

	/**
	 * Only include transactions with this address in one of the transaction fields. 
	 */
	public LookupAssetTransactions setAddress(String address) {
		this.address = address;
		return this;
	}

	/**
	 * Combine with the address parameter to define what type of address to search for. 
	 */
	public LookupAssetTransactions setAddressRole(String addressRole) {
		this.addressRole = addressRole;
		return this;
	}

	/**
	 * Include results after the given time. Must be an RFC 3339 formatted string. 
	 */
	public LookupAssetTransactions setAfterTime(String afterTime) {
		this.afterTime = afterTime;
		return this;
	}

	/**
	 * Include results before the given time. Must be an RFC 3339 formatted string. 
	 */
	public LookupAssetTransactions setBeforeTime(String beforeTime) {
		this.beforeTime = beforeTime;
		return this;
	}

	/**
	 * Results should have an amount greater than this value. MicroAlgos are the 
	 * default currency unless an asset-id is provided, in which case the asset will be 
	 * used. 
	 */
	public LookupAssetTransactions setCurrencyGreaterThan(Long currencyGreaterThan) {
		this.currencyGreaterThan = currencyGreaterThan;
		return this;
	}

	/**
	 * Results should have an amount less than this value. MicroAlgos are the default 
	 * currency unless an asset-id is provided, in which case the asset will be used. 
	 */
	public LookupAssetTransactions setCurrencyLessThan(Long currencyLessThan) {
		this.currencyLessThan = currencyLessThan;
		return this;
	}

	/**
	 * Combine with address and address-role parameters to define what type of address 
	 * to search for. The close to fields are normally treated as a receiver, if you 
	 * would like to exclude them set this parameter to true. 
	 */
	public LookupAssetTransactions setExcludeCloseTo(Boolean excludeCloseTo) {
		this.excludeCloseTo = excludeCloseTo;
		return this;
	}

	/**
	 * Maximum number of results to return. 
	 */
	public LookupAssetTransactions setLimit(Long limit) {
		this.limit = limit;
		return this;
	}

	/**
	 * Include results at or before the specified max-round. 
	 */
	public LookupAssetTransactions setMaxRound(Long maxRound) {
		this.maxRound = maxRound;
		return this;
	}

	/**
	 * Include results at or after the specified min-round. 
	 */
	public LookupAssetTransactions setMinRound(Long minRound) {
		this.minRound = minRound;
		return this;
	}

	/**
	 * The next page of results. Use the next token provided by the previous results. 
	 */
	public LookupAssetTransactions setNext(String next) {
		this.next = next;
		return this;
	}

	/**
	 * Specifies a prefix which must be contained in the note field. 
	 */
	public LookupAssetTransactions setNotePrefix(String notePrefix) {
		this.notePrefix = notePrefix;
		return this;
	}

	/**
	 * Include results for the specified round. 
	 */
	public LookupAssetTransactions setRound(Long round) {
		this.round = round;
		return this;
	}

	/**
	 * SigType filters just results using the specified type of signature: * sig - 
	 * Standard * msig - MultiSig * lsig - LogicSig 
	 */
	public LookupAssetTransactions setSigType(String sigType) {
		this.sigType = sigType;
		return this;
	}

	/**
	 * Lookup the specific transaction by ID. 
	 */
	public LookupAssetTransactions setTxId(String txId) {
		this.txId = txId;
		return this;
	}
	public LookupAssetTransactions setTxType(String txType) {
		this.txType = txType;
		return this;
	}

	@Override
	public Response<TransactionsResponse> execute() throws Exception {
		Response<TransactionsResponse> resp = baseExecute();
		resp.setValueType(TransactionsResponse.class);
		return resp;
	}
	public QueryData getRequestString() {
		QueryData qd = new QueryData();
		if (this.address != null) {
			qd.addQuery("address", String.valueOf(address));
		}
		if (this.addressRole != null) {
			qd.addQuery("addressRole", String.valueOf(addressRole));
		}
		if (this.afterTime != null) {
			qd.addQuery("afterTime", String.valueOf(afterTime));
		}
		if (this.beforeTime != null) {
			qd.addQuery("beforeTime", String.valueOf(beforeTime));
		}
		if (this.currencyGreaterThan != null) {
			qd.addQuery("currencyGreaterThan", String.valueOf(currencyGreaterThan));
		}
		if (this.currencyLessThan != null) {
			qd.addQuery("currencyLessThan", String.valueOf(currencyLessThan));
		}
		if (this.excludeCloseTo != null) {
			qd.addQuery("excludeCloseTo", String.valueOf(excludeCloseTo));
		}
		if (this.limit != null) {
			qd.addQuery("limit", String.valueOf(limit));
		}
		if (this.maxRound != null) {
			qd.addQuery("maxRound", String.valueOf(maxRound));
		}
		if (this.minRound != null) {
			qd.addQuery("minRound", String.valueOf(minRound));
		}
		if (this.next != null) {
			qd.addQuery("next", String.valueOf(next));
		}
		if (this.notePrefix != null) {
			qd.addQuery("notePrefix", String.valueOf(notePrefix));
		}
		if (this.round != null) {
			qd.addQuery("round", String.valueOf(round));
		}
		if (this.sigType != null) {
			qd.addQuery("sigType", String.valueOf(sigType));
		}
		if (this.txId != null) {
			qd.addQuery("txId", String.valueOf(txId));
		}
		if (this.txType != null) {
			qd.addQuery("txType", String.valueOf(txType));
		}
		qd.addPathSegment(String.valueOf("assets"));
		qd.addPathSegment(String.valueOf(assetId));
		qd.addPathSegment(String.valueOf("transactions"));

		return qd;
	}
}

File: generator/go-sdk/indexer/lookupAssetTransactions.go
package indexer

import (
	"context"
	"encoding/base64"
	"fmt"
	"time"

	"github.com/algorand/go-algorand-sdk/client/v2/common"
	"github.com/algorand/go-algorand-sdk/client/v2/common/models"
	"github.com/algorand/go-algorand-sdk/types"
)

// LookupAssetTransactions /v2/assets/{asset-id}/transactions
// Lookup transactions for an asset.
type LookupAssetTransactions struct {
	c       *Client
	p       models.LookupAssetTransactionsParams
	assetId uint64
}

// Address only include transactions with this address in one of the transaction
// fields.
func (s *LookupAssetTransactions) Address(address types.Address) *LookupAssetTransactions {
	s.p.Address = address.String()
	return s
}

// AddressRole combine with the address parameter to define what type of address to
// search for.
func (s *LookupAssetTransactions) AddressRole(addressRole string) *LookupAssetTransactions {
	s.p.AddressRole = addressRole
	return s
}

// AfterTime include results after the given time. Must be an RFC 3339 formatted
// string.
func (s *LookupAssetTransactions) AfterTime(afterTime time.Time) *LookupAssetTransactions {
	s.p.AfterTime = afterTime.Format(time.RFC3339)
	return s
}

// BeforeTime include results before the given time. Must be an RFC 3339 formatted
// string.
func (s *LookupAssetTransactions) BeforeTime(beforeTime time.Time) *LookupAssetTransactions {
	s.p.BeforeTime = beforeTime.Format(time.RFC3339)
	return s
}

// CurrencyGreaterThan results should have an amount greater than this value.
// MicroAlgos are the default currency unless an asset-id is provided, in which
// case the asset will be used.
func (s *LookupAssetTransactions) CurrencyGreaterThan(currencyGreaterThan uint64) *LookupAssetTransactions {
	s.p.CurrencyGreaterThan = currencyGreaterThan
	return s
}

// CurrencyLessThan results should have an amount less than this value. MicroAlgos
// are the default currency unless an asset-id is provided, in which case the asset
// will be used.
func (s *LookupAssetTransactions) CurrencyLessThan(currencyLessThan uint64) *LookupAssetTransactions {
	s.p.CurrencyLessThan = currencyLessThan
	return s
}

// ExcludeCloseTo combine with address and address-role parameters to define what
// type of address to search for. The close to fields are normally treated as a
// receiver, if you would like to exclude them set this parameter to true.
func (s *LookupAssetTransactions) ExcludeCloseTo(excludeCloseTo bool) *LookupAssetTransactions {
	s.p.ExcludeCloseTo = excludeCloseTo
	return s
}

// Limit maximum number of results to return.
func (s *LookupAssetTransactions) Limit(limit uint64) *LookupAssetTransactions {
	s.p.Limit = limit
	return s
}

// MaxRound include results at or before the specified max-round.
func (s *LookupAssetTransactions) MaxRound(maxRound uint64) *LookupAssetTransactions {
	s.p.MaxRound = maxRound
	return s
}

// MinRound include results at or after the specified min-round.
func (s *LookupAssetTransactions) MinRound(minRound uint64) *LookupAssetTransactions {
	s.p.MinRound = minRound
	return s
}

// Next the next page of results. Use the next token provided by the previous
// results.
func (s *LookupAssetTransactions) Next(next string) *LookupAssetTransactions {
	s.p.Next = next
	return s
}

// NotePrefix specifies a prefix which must be contained in the note field.
func (s *LookupAssetTransactions) NotePrefix(notePrefix []byte) *LookupAssetTransactions {
	s.p.NotePrefix = base64.StdEncoding.EncodeToString(notePrefix)
	return s
}

// RekeyTo include results which include the rekey-to field.
func (s *LookupAssetTransactions) RekeyTo(rekeyTo bool) *LookupAssetTransactions {
	s.p.RekeyTo = rekeyTo
	return s
}

// Round include results for the specified round.
func (s *LookupAssetTransactions) Round(round uint64) *LookupAssetTransactions {
	s.p.Round = round
	return s
}

// SigType filters just results using the specified type of signature:
// * sig - Standard
// * msig - MultiSig
// * lsig - LogicSig
func (s *LookupAssetTransactions) SigType(sigType string) *LookupAssetTransactions {
	s.p.SigType = sigType
	return s
}

func (s *LookupAssetTransactions) TxType(txType string) *LookupAssetTransactions {
	s.p.TxType = txType
	return s
}

// TxID lookup the specific transaction by ID.
func (s *LookupAssetTransactions) TxID(txID string) *LookupAssetTransactions {
	s.p.TxID = txID
	return s
}

// Do performs HTTP request
func (s *LookupAssetTransactions) Do(ctx context.Context,
	headers ...*common.Header) (response models.TransactionsResponse, err error) {
	err = s.c.get(ctx, &response,
		fmt.Sprintf("/v2/assets/%d/transactions", s.assetId), s.p, headers)
	return
}
